### PR TITLE
4337 module formal verification

### PR DESCRIPTION
**Pull request by akshay-ap**
This PR is related to #142

Changes in PR:
- Setup spec files, and configuration files for Formal Verification using Certora
- Setup harnesses contracts
- Add GitHub workflow file

Rules Implemented:

- onlyEntryPointCallable
This rule verifies that methods related to 4337 are only callable by the entrypoint contract: `validateUserOp(..)`, `executeUserOp(...)` and `executeUserOpWithErrorString(...)`

- checkSignaturesIsCalledIfValidateUserOpSucceeds
This rule verifies that `checkSignatures(...)` is always called `validateUserOp(...)` function is invoked

- signatureTimestampsPresentInValidationData
This rule verifies that `validAfter` and `validUntil` are present in the first 12 bytes of the validation data returned by the `validateUserOp(...)` function

- validationDataLastBitZeroIfCheckSignaturesSucceeds
This rule verifies that validation data returned by the `validateUserOp(...)` contains the last bit `0` if signature validation succeeds

2. ExecTransactionFromModule.spec 
This rule should verify that `safe.execTransactionFromModule(...)` is only called by `validateUserOp(..)`, `executeUserOp(...)` and `executeUserOpWithErrorString(...)`

3. PayMissingFunds.spec
This rule should verify that the Module sends only `missingAccountFunds` from Safe to EntryPoint contract

4. ValidationDataLastBitOne.spec
This rule should verify that validation data returned by the `validateUserOp(...)` contains the last bit `1` if signature validation fails

### CODE REVIEW

File: 4337/certora/conf/ExecTransactionFromModule.conf [hunk 1]
```diff
@@ -0,0 +1,24 @@
+{
+    "files": [
+        "contracts/Safe4337Module.sol",
+        "contracts/test/certora/Account.sol:Account2",
+        "node_modules/@safe-global/safe-contracts/contracts/Safe.sol",
+
```
Comment (nlordell) [inline 4337/certora/conf/ExecTransactionFromModule.conf:6]:
```suggestion
```

File: 4337/certora/scripts/verify.sh [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+# certoraRun certora/conf/ExecTransactionFromModule.conf \
+#     "${params[@]}" \
+#     --msg "Safe4337Module $*" \
+#     "$@"
+
+# certoraRun certora/conf/ValidationDataLastBitOne.conf \
+#     "${params[@]}" \
+#     --msg "Safe4337Module $*" \
+#     "$@"
```
Comment (nlordell) [inline 4337/certora/scripts/verify.sh:27]: Should these be commented out?

File: 4337/contracts/test/certora/Account.sol [hunk 3]
```diff
@@ -0,0 +1,140 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity >=0.8.0;
+import {Safe} from "@safe-global/safe-contracts/contracts/Safe.sol";
+import {Enum} from "@safe-global/safe-contracts/contracts/common/Enum.sol";
+
+contract Account is Safe {
```
Comment (nlordell) [inline 4337/contracts/test/certora/Account.sol:6]: Maybe `Account1` for consistency?

File: 4337/contracts/test/certora/Account.sol [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        uint256 value,
+        bytes memory data,
+        Enum.Operation operation
+    ) public override returns (bool success) {
+        execTransactionFromModuleCalled = true;
+        super.execTransactionFromModule(to, value, data, operation);
+    }
+}
+
+contract Account4 {
```
Comment (nlordell) [inline 4337/contracts/test/certora/Account.sol:87]: uber-nit (so feel free to ignore): the contracts are out of order (1 -> 2 -> 4 -> 3).

**Review by nlordell — APPROVED** [review]
In general these rules look good, although I don't feel well equipped to review them in a meaningful way :/.

File: .github/workflows/certora.yml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+              run: pip install -Iv certora-cli
+
+            - name: Install solc
+              run: |
+                  wget https://github.com/ethereum/solidity/releases/download/v0.8.23/solc-static-linux
+                  chmod +x solc-static-linux
+                  sudo mv solc-static-linux /usr/local/bin/solc8.23
+
+            - name: Install dependencies
+              run: yarn --frozen-lockfile
```
Comment (mmv08) [inline .github/workflows/certora.yml:44]: the 4337 module uses npm

File: 4337/certora/conf/PayMissingFunds.conf [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    "loop_iter": "3",
+    "msg": "Safe4337Module: PayForMissingFunds Ruleset",
+    "optimistic_hashing": true,
+    "optimistic_loop": true,
+    "process": "emv",
+    "prover_args": [
+        " -optimisticFallback true -s z3 -copyLoopUnroll 5 -mediumTimeout 5 -depth 30"
+    ],
+    "rule_sanity": "advanced",
+    // "solc": "solc8.23",
```
Comment (mmv08) [inline 4337/certora/conf/PayMissingFunds.conf:17]: why's this commented out? (same question applies for all conf files)

File: 4337/certora/scripts/verify.sh (hunk 2 above)
Comment (mmv08) [reply 4337/certora/scripts/verify.sh:27]: The PR description says that they're wip, so I guess so. Would be good to add a comment tho

File: 4337/certora/specs/ExecTransactionFromModule.spec [hunk 7]
```diff
@@ -0,0 +1,60 @@
+using Account2 as safeContract;
+
+methods {
```
Comment (mmv08) [inline 4337/certora/specs/ExecTransactionFromModule.spec:3]: I know that this is a WIP spec file, but I'd only include methods that are used by the rule here

File: 4337/certora/specs/ExecTransactionFromModule.spec [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        uint48 validAfter,
+        uint48 validUntil
+    ) external returns(bytes32) envfree => CONSTANT;
+}
+
+ghost bool execTransactionFromModuleCalled;
+
+function ExecTxCalled() returns bool {
+    execTransactionFromModuleCalled = true;
+    return true;
```
Comment (mmv08) [inline 4337/certora/specs/ExecTransactionFromModule.spec:43]: is return value needed here? doesn't seem like it's used anywhere

File: 4337/contracts/test/certora/Account.sol [hunk 9]
```diff
@@ -0,0 +1,140 @@
+// SPDX-License-Identifier: LGPL-3.0-only
```
Comment (mmv08) [inline 4337/contracts/test/certora/Account.sol:1]: Wdyt about including this in the certora folder directly, similar to how it's done in the safe-contracts repo?

File: 4337/certora/specs/Safe4337Module.spec [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+        uint48 validUntil
+    ) external returns(bytes32) envfree => CONSTANT;
+}
+ghost ERC2771MessageSender() returns address;
+
+ghost bool checkSignaturesCalled;
+
+function checkSignaturesFunctionCalled() returns bool{
+    checkSignaturesCalled = true;
+    return true;
```
Comment (mmv08) [inline 4337/certora/specs/Safe4337Module.spec:45]: same question about the return value, is it needed?

File: 4337/certora/specs/Safe4337Module.spec [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+        uint256 missingAccountFunds) {
+    env e;
+    uint48 validAfter;
+    uint48 validUntil;
+    require validAfter == safeContract.getValidAfterTimestamp(userOp.signature);
+    require validUntil == safeContract.getValidUntilTimestamp(userOp.signature);
+    checkSignaturesCalled = false;
+
+    uint256 validationData = validateUserOp@withrevert(e, userOp, dummyData, missingAccountFunds);
+    assert !lastReverted => checkSignaturesCalled, "transaction executed without valid signatures";
```
Comment (mmv08) [inline 4337/certora/specs/Safe4337Module.spec:72]: the error message seems a tad bit incorrect here, i'd say something like "validation passed without checking signatures"

**Comment by mmv08** [comment]
@nlordell The PR is available for review.

What I've done:
- All the rules are done now
- The amount of harness accounts is down from 4 to 2
- Minor cleanups in the methods block

File: .gitignore [hunk 12]
```diff
@@ -14,3 +14,4 @@ env/
 .env
 .history
 coverage*
+.certora_internal
```
Comment (nlordell) [inline .gitignore:17]: nit: missing newline at EOF

File: 4337/certora/conf/TransactionExecutionMethods.conf [hunk 13]
```diff
@@ -0,0 +1,22 @@
+{
+    "files": [
+        "contracts/Safe4337Module.sol",
+
```
Comment (nlordell) [inline 4337/certora/conf/TransactionExecutionMethods.conf:4]:
```suggestion
```

**Review by nlordell — APPROVED** [review]
This feels a bit like a rubber-stamp approval :sweat_smile:. I briefly looked through the rules and they look correct AFAICT with my limited knowledge.

**Comment by mmv08** [comment]
> This feels a bit like a rubber-stamp approval 😅. I briefly looked through the rules and they look correct AFAICT with my limited knowledge.

I think it looks OK. I asked the Certora team for a quick check and if there are any apparent rules that we haven't come up with.

**Comment by nlordell** [comment]
Good enough for me.

**Pull request merged**